service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java

```diff
@@ -0,0 +1,151 @@
+import io.smallrye.stork.ServiceInstance;
+import io.smallrye.stork.config.ServiceDiscoveryConfig;
+import io.smallrye.stork.spi.ServiceInstanceIds;
+import io.vertx.core.Vertx;
+
+public class KubernetesServiceDiscovery implements ServiceDiscovery {
+
+    private final KubernetesClient client;
+    private final String serviceName;
+    private boolean allNamespaces = false;
+    private String namespace;
+    private final Vertx vertx;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KubernetesServiceDiscovery.class);
+
+    public KubernetesServiceDiscovery(String serviceName, ServiceDiscoveryConfig config, Vertx vertx) {
+        Config base = Config.autoConfigure(null);
+        this.serviceName = serviceName;
+        Map<String, String> parameters = config.parameters();
+        String masterUrl = parameters != null ? parameters.get("k8s-host") : base.getMasterUrl();
+        if (masterUrl == null) {
+            masterUrl = base.getMasterUrl();
+        }
+        namespace = parameters != null ? parameters.get("k8s-namespace") : base.getNamespace();
+        if (namespace != null && namespace.equalsIgnoreCase("all")) {
+            allNamespaces = true;
+        }
+        Config properties = new ConfigBuilder(base)
+                .withMasterUrl(masterUrl)
+                .withNamespace(namespace).build();
+        client = new DefaultKubernetesClient(properties);
+        this.vertx = vertx;
+    }
+
+    public Uni<List<ServiceInstance>> blockingGetServiceInstances() {
+        List<Endpoints> endpoints = allNamespaces
+                ? client.endpoints().inAnyNamespace().withField("metadata.name", serviceName).list()
+                        .getItems()
+                : client.endpoints().inNamespace(namespace).withField("metadata.name", serviceName).list().getItems();
+        Uni<List<ServiceInstance>> serviceEntryList = Uni.createFrom().item(map(endpoints));
+        return serviceEntryList; // TODO: logging
+    }
+
+    public Uni<List<ServiceInstance>> getServiceInstances() {
+        Uni<List<Endpoints>> endpointsUni = Uni.createFrom().emitter(
+                emitter -> {
+                    List<Endpoints> endpoints = new ArrayList<>();
+                    vertx.executeBlocking(future -> {
+                        if (allNamespaces) {
+                            endpoints.addAll(client.endpoints().inAnyNamespace().withField("metadata.name", serviceName).list()
+                                    .getItems());
+                        } else {
+                            endpoints.addAll(
+                                    client.endpoints().inNamespace(namespace).withField("metadata.name", serviceName).list()
+                                            .getItems());
+                        }
+                        future.complete();
+                    }, result -> {
+                        if (result.succeeded()) {
+                            emitter.complete(endpoints);
```

Comment:
```suggestion
                    vertx.executeBlocking(future -> {
                        List<Endpoints> endpoints = new ArrayList<>();
                        if (allNamespaces) {
                            endpoints.addAll(client.endpoints().inAnyNamespace().withField("metadata.name", serviceName).list()
                                    .getItems());
                        } else {
                            endpoints.addAll(
                                    client.endpoints().inNamespace(namespace).withField("metadata.name", serviceName).list()
                                            .getItems());
                        }
                        future.complete(endpoints);
                    }, result -> {
                        if (result.succeeded()) {
                            emitter.complete(result.result());
```

I haven't checked it but I think it should be sth like ^. Otherwise it may be not thread safe because the code in `executeBlocking` is executed on another thread than the continuation